python/lsst/pipe/base/pipelineIR.py

```diff
@@ -637,6 +668,28 @@ def _read_labeled_subsets(self, loaded_yaml: dict[str, Any]) -> None:
         for key, value in loaded_subsets.items():
             self.labeled_subsets[key] = LabeledSubset.from_primitives(key, value)
 
+    def _read_step_declaration(self, loaded_yaml: dict[str, Any]) -> None:
+        """Process the steps portion of the loaded yaml document
+
+        Steps are subsets that are declared to be normal parts of the overall
+        processing of the pipeline. Not all subsets need to be a step, as they
+        can exist for certain targeted processing, such as debugging.
+
+        Parameters
+        ----------
+        loaded_yaml: `MutableMapping`
+            A dictionary which matches the structure that would be produced
+            by a yaml reader which parses a pipeline definition document
+        """
+        loaded_steps = loaded_yaml.pop("steps", [])
+        self.steps: set[StepIR] = set()
```

Comment:
I think you can avoid all of the problems in `StepIR.__eq__` and `__hash__` by making this a `dict[str, StepIR]` or even a `dict[str, list[str]]` (and dropping `StepIR`).  It makes the IR slightly more different from the YAML form, but that seems like the lesser evil.

In fact, weren't we planning to make this a `list` in the YAML in order to validate the order, and then we decided not to validate the order because it was incompatible with importing steps?  So maybe this should be a `dict[str, list[str]]` in that YAML, too, after all?  That actually does preserve order anyway, which is nice for round-tripping even if we don't validate order.